src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -737,6 +737,22 @@ int RGWRados::get_max_chunk_size(const rgw_placement_rule& placement_rule, const
   return get_max_chunk_size(pool, max_chunk_size, dpp, palignment);
 }
 
+void add_datalog_entries(const DoutPrefixProvider* dpp,
+                         RGWDataChangesLog* datalog,
+                         const RGWBucketInfo& bucket_info,
+                         std::vector<uint64_t> shards,
+                         optional_yield y)
+{
+  const auto& logs = bucket_info.layout.logs;
+  if (logs.empty()) {
+    return;
+  }
+
+  for (auto shard_id : shards) {
+    // int r = datalog->add_entry_aio(dpp, bucket_info, logs.back(), shard_id, c, y);
```

Comment:
@cbodley I still need to construct a completion object and pass it into add_entry_aio() here
does this need to be a coroutine function where I can call create_completion_notifier()?

Reply:
@adamemerson we'd like your opinion on this. i need to construct and pass in a completion object here, and I think we'd want to get its return value if there is an error.  is it required to put this in a coroutine like you've done for datalog trim_entries().

Reply:
If you want to act on the error rather than simply returning it to the caller, you'll need a state machine like GenTrim.

If you just want to return an error code to the caller and you're done, you should be able to pass the caller's AioCompletion in.

Reply:
thanks @adamemerson.
it looks like we don't care about what happens to the bilog suggested updates. we seem to simply fire them off. I don't think we need to care about handling datalog op failures either @cbodley